docs/engineering/architecture/index.md

```diff
@@ -0,0 +1,138 @@
+
+1. The web service uses the `user` and `database` (the user is picked up from the request object) to acquire a connection string.
+
+1. Using that connection, the web service calls a PL/pgSQL function installed on the User DB called `get_schema_table_details` to gather 
+
+    - `name`
+    - `description`
+    - `has_dependents`
+    - `columns`
+        - `name`
+        - `type`
+        - `type_options`
+    - `preview_settings` -- describes how we should show each table's rows when it's linked to by a foreign key
+
+1. Using the returned info, the web service filters a `TableMetadata` model based on returned `oid`s, and gathers
+
+    - `import_verified`
+    - `is_temp`
+    - `import_target_id`
+    - `data_files`
+    - `column_order` -- describes the order in which columns should be displayed
+
+1. Using the same returned info, filter a `ColumnMetadata` model based on returned `oid, attnum` pairs to gather (for each column)
+
+    - `display_options`
+
+We then join all of this together, and return it as a response from the API.
+    
+The fundamental difference is that in the current version, we use foreign keys between Django models to find tables for the schema, then columns for each table. Then all queries on the User DB are initiated by functions on these model instances. In the new version, we instead run a query on the User DB to gather all relevant table and column info available on that DB, then enrich that data with metadata stored in non-foreign-key-linked metadata models in the Django DB.
+
+## Introduction to relevant layers
+
+The layers introduced here will be discussed in more detail in other sections.
+
+### User database
+
+For each API call, there should be an identifiable DB function that performs all User Database operations needed to satisfy that call. For example,
+
+- Calling the function `get_tables(database, schema)` should result in a call to some function `get_tables(sch_oid oid)` on the `database`.
+
+To achieve this, we will install the following on the user database(s)
+
+- Some custom Mathesar types.
+- Some tables holding metadata that needs to be kept near the tables.
+- A set of functions that provide the bulk of Mathesar's back end logic and functionality.
+
+### Python `db` library
+
+This library should mostly serve to provide thin wrapper functions around the User DB layer functions. These functions should take parameters from requests (never request objects themselves) and engines, and then call underlying DB functions. They should then pass the results up towards the API
+
+### Web service
+
+This service should provide an JSON-RPC API for use by the front end. When an API function is called, the service should:
+
+- Grab an appropriate engine from the `DatabaseConnection` model by combining the user associated with the request with the `database`.
```

Comment:
What `DatabaseConnection` model? Could you describe the new model structure somewhere and compare it to the current structure?

Reply:
Never mind, found the models file, but can you link it somewhere from here?

Reply:
Okay I found the models file but I am finding it very hard to follow with the ad-hoc notes, can we do some sort of comparison of old structure vs. new structure?